### DISCUSSION TITLE
Quarto: Locals, globals, dynamic text

### DISCUSSION THREAD

**Discussion by friosavila** [hugetim/nbstata#11] [General]
I tried to look into the documentation but couldn't find information about the following.
Is it possible to do inline coding?
for example something like this [inline](https://grodri.github.io/markstat/documentation).
So that if I type `stata e(r2)` (or similar) to display the desired outcome? (Print R2 of a regression, in line)
Thank you

**Comment by hugetim** [comment]
I think this may again be more of a Quarto question. (I'm guessing that you are doing something like [this](https://quarto.org/docs/computations/python.html) but with `jupyter: nbstata` in the header?) `nbstata` basically provides two well-defined interfaces:
1. A Stata [Jupyter kernel](https://docs.jupyter.org/en/latest/projects/kernels.html#kernels)
2. A Python API to assist using Stata from within Python code (currently, this is documented only in the [Developer Docs](https://hugetim.github.io/nbstata/dev_docs_index.html))

If you are using an external system (other than Jupyter Lab) like Quarto that plugs in to one of these interfaces (in Quarto's case, the kernel), your questions/doc searches are most likely better directed there.

**Comment by friosavila** [comment]
Thank you for the information
Best wishes

**Comment by hugetim** [comment]
I'm really interested in learning more from your experience of using nbstata with Quarto (or other platforms), though. Anything you are willing to share would be most welcome.

**Comment by friosavila** [comment]
Certainly!
I recently started working using Quarto. So I'm mostly doing some work with
Stata. NOw I'm trying to integrate the examples with Stata code and Output.
This will save me time, as opposed to running all examples outside quarto,
and just copy-pasting the outcomes.
Im also planning to write a "summary statistics paper". I can certainly
share that with you when that is done.

For now, in regards to my Local question, i decided to do it internally,
using something like this:

```{stata}
*| output: asis
*| echo  : false
qui:sysuse auto, clear
qui:regress price mpg
display "Based on the model, when a car is more fuel efficient, the cost
decreases in " %5.1f `=_b[mpg]' " dollars."
```
This should work since only few of the dynamic parameters in the paper
require explicit explanation.

Thank you
Fernando

**Reply by hugetim** [reply]
Ok, I see. And using that `*| output: asis` directive, you can even output Markdown and have that formatted correctly in the Quarto output. For example, you could add emphasis to a couple words like this:
```stata
display "Based on the model, when a car is *more* fuel efficient, the cost
*decreases* in " %5.1f `=_b[mpg]' " dollars."
```
And the Quarto output shows those words in italics. (I just tried it. I'm up and running with Quarto now..)

**Comment by rdimond** [comment]
A colleague pointed me to a thread on Statalist that led me here. I'm getting great results using nbstata to write Stata Jupyter Notebooks in JupyterLab and then rendering them using Quarto. 

Article example: https://sscc.wisc.edu/sscc/pubs/stata_dates/
Web book example: https://sscc.wisc.edu/sscc/pubs/stata_tables/

Setting up the site formatting in Quarto took some figuring out, but not too bad. I'm still figuring out best ways to use it.

I switched to nbstata because it has the echo behavior I want. I did have some trouble with JuptyerLab on my laptop after installing nbstata and Stata syntax highlighting, but since they just work on our Linux servers I haven't investigated. (I can start JupyterLab on a Linux server and point a local browser to it, so it's just like running it on my laptop anyway.)

Thank you Tim!

Russell Dimond

**Reply by hugetim** [reply]
That's awesome to hear! I really appreciate you sharing.

Anything you can pass on about what you learned setting up Quarto to work with `nbstata` is most welcome as I'm looking to try that soon myself. (I'm also curious about the laptop troubles but no worries if you've moved on from that.)

**Reply by rdimond** [reply]
Making Quarto work with nbstata took no specific work at all. We use the exact same Quarto config files with Python (https://sscc.wisc.edu/sscc/pubs/dwp/) and R (https://www.sscc.wisc.edu/sscc/pubs/slurm/R_slurm_loops.html). It also doesn't matter if the source is a Jupyter Notebook or an RMarkdown file written in RStudio. Looks like Posit wanting to be more universal and less R-centric is more than just branding!

As for the laptop troubles, the first thing I need to do is blow everything away and see if it was installing nbstata or the syntax highlighting that caused the problem--I suspect it was the latter. It also won't surprise me if the problem goes away on a reinstall.

**Reply by hugetim** [reply]
That would make sense. The syntax highlighting extension was much more finicky to get working for me (as I know so little JavaScript).

When starting from a Jupyter Notebook with nbstata, is it an issue that the Quarto directives (starting with `#|`) aren't valid Stata comments? Would it be helpful if I made nbstata ignore those lines?

**Reply by hugetim** [reply]
Nevermind, I see from @friosavila's example above (and now below 👀) that you can just use `*|` for the directives. Amazing. I guess it learns from the jupyter kernel spec what the line comment character is.

**Reply by hugetim** [reply]
@rdimond I don't know how much trouble it would be for you to do this after the fact, but it might help Googlers find `nbstata` if you were able to correct the spelling of "nbstata" in the [Stata dates](https://sscc.wisc.edu/sscc/pubs/stata_dates/) article (from "`nbstat` kernel"). And even better would be to link to https://hugetim.github.io/nbstata/ from that page.

(I just happened to notice this as I was sending that page to someone as an example of using nbstata with Quarto.)

**Comment by friosavila** [comment]
I saw that.
Because Stata uses * for comments, you also have to use * to provide
directives in Quarto

for example:
```{stata}
*| echo: false
*| output: asis
qui:reg price mpg foreign
est sto m1
esttab m1, md nonum
```
This is what I used to create the toy example I posted on Statalist

**Reply by hugetim** [reply]
[Link to referenced Statalist post with example nbstata Quarto input and output files](https://www.statalist.org/forums/forum/general-stata-discussion/general/1703835-ado-files-and-literate-programming?p=1705255#post1705255)

**Comment by hugetim** [comment]
Quarto is blowing my mind. It seemed impenetrable as I browsed randomly through the reference docs to find answers to particular questions. But then I finally broke down and...worked through the tutorials. 🌩💡 

Working through the tutorials will be a bigger challenge for Stata users without experience in any of the tutorial reference languages, so I'd like to add a Stata Quarto tutorial to the docs here soon.

**Comment by friosavila** [comment]
That is an excellent idea!

**Comment by friosavila** [comment]
Are you thinking just using the same design but with Stata as the main
language?

**Reply by hugetim** [reply]
Yes, that was what I was thinking. We could either copy and adapt the qmd source of the web pages...or the less ambitious option would be to just create sample notebooks and .qmd files that folks could reference and play with as they work through the Python or R tutorial.

**Comment by friosavila** [comment]
Started to look into it
It will definitely need some tweaking for rare cases
but, for the examples they provide (computation), perhaps the following
will be enough

```{stata}
mata:a = rowshape(1::15, 5)
mata:a
```

```{stata}
*| label: fig-limits
*| fig-cap: "Citations vs Fines"
qui:webuse dui, clear
two scatter citations fines || lpolyci citations fines, xsize(10) ysize(5)

```

```{stata}
*| label: fig-gapminder
*| fig-cap: Life Expectancy and GDP
*| fig-subcap:
*|   - "Gapminder: 1957"
*|   - "Gapminder: 2007"
*| layout-ncol: 2
*| column: page
*| echo: false
qui:sysuse auto, clear

scatter price mpg if foreign==1, name(m1, replace)
scatter price mpg if foreign==0, name(m2, replace)
```

NOTE: Mata output is odd. Because every one comes up in the screen. Kind of
how it is when "doing" a dofile
But not sure what is the better approach

For figures, The sizes and resolutions seem to be better defined using
xsize and ysize

and for multiple figures, it seems necessary to "save" graphs in memory in
order to use the layout for multiple figures.
F

**Reply by hugetim** [reply]
Cool, thanks for working on this!

Your observations about figures are as I'd expect.

And yeah, mata commands are echoed in nbstata because I couldn't figure out a way for them not to. I don't use mata myself, so I haven't tried too hard yet. But there might be [a way](https://github.com/ticoneva/pystata-kernel/issues/4#issuecomment-1214284692) if quietly/noisily prefixes are allowed in mata...

For this purpose, though, we could set mata aside and just do something else in that space. Or we could just do a Stata-python block, unless that would be confusing for people:
```
```{stata}
python:
    import numpy as np
    a = np.arange(15).reshape(3, 5)
    print(a)
end
```

**Reply by hugetim** [reply]
It's probably better to just do a common Stata thing like `summarize`.

**Reply by friosavila** [reply]
So, I added a new example here:
https://friosavila.github.io/playingwithstata/quarto/dynstat
https://github.com/friosavila/playingwithstata/tree/gh-pages/quarto

**Comment by rdimond** [comment]
Thanks for catching that—it’s been corrected and a link added.

Russell

**Comment by hugetim** [comment]
New Statalist Quarto thread: https://www.statalist.org/forums/forum/general-stata-discussion/general/1765340-literate-programming-and-cross-reference-of-formatted-tables
